### DISCUSSION TITLE
How to re-layout the graph after deleting some nodes ?

### DISCUSSION THREAD

**Discussion by Nazeh-Taha** [xyflow/xyflow#1312] [Q&A]
How to re-layout the graph after deleting some nodes?

```const getLayoutedElements = (els, direction = "BT") => {
    dagreGraph.setGraph({ rankdir: direction });
    els.forEach((el) => {
      if (el.type !== "strategynode") {
        if (isNode(el)) {
          dagreGraph.setNode(el.id, {
            width: findWidth(el),
            height: 130,
          });
        } else {
          dagreGraph.setEdge(el.source, el.target);
        }
      }
    })

    dagre.layout(dagreGraph);

    return els.map((el) => {
      if (el.type !== "strategynode") {
        if (isNode(el)) {
          const nodeWithPosition = dagreGraph.node(el.id);
          el.position = {
            x: nodeWithPosition.x - findWidth(el) / 2 + Math.random() / 1000,
            y: nodeWithPosition.y - nodeHeight / 2,
          };
        }
      }
      return el;
    });
  };
```

` const onElementsRemove = (elementsToRemove) => {
    setElements((els) => removeElements(elementsToRemove, els));
    getLayoutedElements(elements);
  };`

**Everything is good:**

![1](https://user-images.githubusercontent.com/54739625/123546432-25c0e780-d765-11eb-9ec6-3af933604001.jpg)

**After deleting nodes:**

![2](https://user-images.githubusercontent.com/54739625/123546454-3a9d7b00-d765-11eb-9596-196f055b43f7.jpg)

**After adding the same nodes again:**

![3](https://user-images.githubusercontent.com/54739625/123546472-53a62c00-d765-11eb-8019-aefd8e49ccdf.jpg)

**Comment by moklick** [comment]
You need to be sure that the elements array is updated when you pass it to the `getLayoutedElements` function.

**Reply by Nazeh-Taha** [reply]
Thanks @moklick for answer, But I sure the array is updated but still have this spaces

**Reply by jayanthbondi** [reply]
@Nazeh-Taha Could you please let me know how you have achieved it?
I'm facing the same issue.

**Reply by Nazeh-Taha** [reply]
@jayanthbondi Sorry, I didn't solve it, there are changes in requirements, so I forced to change this library.
I don't know if it is proper to share the library name here, I will share it for benefits 🙏
[basicprimitives](https://www.basicprimitives.com/)
[npm](https://www.npmjs.com/package/basicprimitivesreact)

**Reply by harini24** [reply]
re- initializing the dagre graph seems to be fixing this issue
`const dagreGraph = new dagre.graphlib.Graph();`
